### PR TITLE
Add support for slicing in h5_read, allow non-contiguous views in read/write

### PR DESCRIPTION
**Pull request by Wentzell**
Allow to read only a slice of the data in file into an array/view by providing a slice tuple e.g.
`h5_read(file, "key", arr, std::tuple{nda::range(2,4,2), 3})`
The array or view to read into should have a rank matching the number of non-trivial (ranged) slice dimensions.

This PR triggered https://github.com/TRIQS/h5/commit/3ddcd36a1e23e5014b4e61c6ce595e7da65b179b which should be included in the review.

### CODE REVIEW

**Comment by Wentzell** [comment]
@parcollet Important note: This will currently not behave as expected for a slice using `range()` to keep all indices for one of the dimensions. The reason here is that `range()` is really the same as `range(0,-1,1)`, in other words an empty index range.
It's in fact a bit misleading that this denotes a 'full slice' in the case of array slicing.

**Comment by Wentzell** [comment]
I will make sure to extend the slicing to allow the use of `range::all_t` however.

File: c++/nda/h5.hpp [hunk 1]
```diff
@@ -121,20 +126,70 @@ namespace nda {
...
+   * The hyperslab will have the same number of dimensions as the length of the slice
+   * The shape will only contain those dimensions of the slice that are not of integer type
+   *
+   * @param slice The slice tuple
+   * @param lengths The dimensions of the underlying dataset
+   * @param is_complex True iff the dataset holds complex values
+   * @tparam TU The type of the slice tuple, containing integers and/or index ranges
+   */
+  template <typename TU>
+  auto hyperslab_and_shape_from_slice(TU const &slice, std::vector<h5::hsize_t> const &lengths, bool is_complex) {
```
Comment (parcollet) [inline c++/nda/h5.hpp]: TU -> why not std::tuple<T...> 
then you have T (no tuple_element). 
Or do you want to allow any tuple-like TU ?

File: c++/nda/h5.hpp [hunk 2]
```diff
@@ -121,20 +126,70 @@ namespace nda {
...
+    auto hsl                   = h5::array_interface::hyperslab(std::tuple_size_v<TU>, is_complex);
+    constexpr auto sliced_rank = []<size_t... Is>(std::index_sequence<Is...>) {
+      return (std::is_same_v<std::tuple_element_t<Is, TU>, nda::range> + ...)
```
Comment (parcollet) [inline c++/nda/h5.hpp]: Count the integer instead, there is only one thing to count
std::tuple_size_v<TU> - [](......) { return (std::integral<std::tuple_element_t<Is, TU>> + ...);}

Comment (parcollet) [reply c++/nda/h5.hpp]: What about `ellipsis` ? Shall we also permit them here ?

Comment (parcollet) [reply c++/nda/h5.hpp]: What about `block` in the hyperslab ? It has no equivalent in nda

File: c++/nda/h5.hpp [hunk 3]
```diff
@@ -121,20 +126,70 @@ namespace nda {
...
+             hsl.count[n] = lengths[n];
+             shape[m++]   = lengths[n];
+           }
+         }(Is, std::get<Is>(slice)),
+         ...);
+    }
+    (std::make_index_sequence<std::tuple_size_v<TU>>{});
+    return std::make_pair(hsl, shape);
+  }
+
```
Comment (parcollet) [inline c++/nda/h5.hpp:162]: Either this or (a matter of taste)
```
auto l = overload
      [&](long i) {
        hsl.offset[n] = ir;
        hsl.count[n] = 1;
      },
      [&](nda::range const &r) {
        hsl.offset[n] = ir.first();
        hsl.stride[n] = ir.step();
        hsl.count[n] = ir.size();
        shape[m++] = ir.size();
      },
      [&](nda::range::all_t) {
        hsl.count[n] = lengths[n];
        shape[m++] = lengths[n];
      });
```
with the usual (should be std:: at some point ? )
```
template <class... Ts> struct overload : Ts... { using Ts::operator()...; };
template <class... Ts> overload(Ts...) -> overload<Ts...>;
```

File: c++/nda/h5.hpp [hunk 4]
```diff
@@ -121,20 +126,70 @@ namespace nda {
...
+  template <MemoryArray A, typename TU>
+  void h5_read(h5::group g, std::string const &name, A &a,
+               TU const &slice) requires(is_regular_or_view_v<A> and is_instantiation_of_v<std::tuple, TU>) {
```
Comment (parcollet) [inline c++/nda/h5.hpp]: `is_instantiation_of_v<std::tuple, TU>`
cf above : here you loose the generality of TU ... 
If it is a tuple, use std::tuple<T...> above.

Comment (parcollet) [reply c++/nda/h5.hpp]: **API**
Should we have a type for this hyper slab ?
hslab{1, 2...}

File: c++/nda/h5.hpp [hunk 5]
```diff
@@ -155,28 +210,47 @@ namespace nda {
...
+        }
       }
-      h5::array_interface::read(g, name, v, lt);
+      h5::array_interface::read(g, name, v, lt, slice_slab);
 
     } else { // generic unknown type to hdf5
+
+      static_assert(!slicing, "Slicing not supported in generic h5_read");
+
       auto g2 = g.open_group(name);
```
Comment (parcollet) [inline c++/nda/h5.hpp:306]: To be discussed.
Should we split the function into an impl and a function handles the slab template ?
Or use slab{} type ? 
In the past, templating h5_read a lot lead to long compilation time at the end ...

**Review by parcollet — CHANGES_REQUESTED** [review]

**Review by parcollet — APPROVED** [review]

File: c++/nda/h5.hpp (hunk 4 above)
Comment (Wentzell) [reply c++/nda/h5.hpp]: - Yes, I have fixed the type to tuple now.
- Why another type for hyperslab? `h5` already provides one.

File: c++/nda/h5.hpp (hunk 3 above)
Comment (Wentzell) [reply c++/nda/h5.hpp:162]: Yes I think both solutions are equally fine.

File: c++/nda/h5.hpp (hunk 2 above)
Comment (Wentzell) [reply c++/nda/h5.hpp]: - Agreed on the integer counting
- Yes, I will look into allowing for ellipsis
- `block` as defined in the h5 hyperslab has no equivalent within the range-based array slicing that we allow here. We should just keep trivial 1x1 blocks.

File: c++/nda/h5.hpp (hunk 1 above)
Comment (Wentzell) [reply c++/nda/h5.hpp]: Agreed. We only want to allow tuples here.

File: c++/nda/h5.hpp (hunk 2 above)
Comment (Wentzell) [reply c++/nda/h5.hpp]: I have extended the implementation to allow for ellipsis

File: c++/nda/h5.hpp (hunk 5 above)
Comment (Wentzell) [reply c++/nda/h5.hpp:306]: The new functionality should not effect the number of instantiations for those cases where we don't slice.
Slicing will not be the common use-case so I don't foresee a problem here.

**Comment by Wentzell** [comment]
I have also added the functionality to use a slice during the `h5_write`. This will assume that the dataset already exists in file. The associated changes in h5 are https://github.com/TRIQS/h5/commit/d6081913d2d42b258dfe362d17933b3d178764d3

**Comment by parcollet** [comment]
Fine. Maybe refactor with less template, but can be done at later stage.

**Review by parcollet — APPROVED** [review]

**Pull request merged**